server/src/main/java/io/druid/segment/realtime/firehose/ChatHandlerResource.java

```diff
@@ -24,30 +24,43 @@
 
 import javax.ws.rs.Path;
 import javax.ws.rs.PathParam;
-import javax.ws.rs.core.Response;
+import javax.ws.rs.core.Context;
+import javax.ws.rs.core.HttpHeaders;
+import java.util.List;
+import java.util.Properties;
 
 @Path("/druid/worker/v1")
 public class ChatHandlerResource
 {
+  public static final String TASK_ID_HEADER = "X-Druid-Task-Id";
+  public static final String TASK_ID_PROPERTY = "druid.indexer.taskId";
+
   private final ChatHandlerProvider handlers;
+  private final String taskId;
 
   @Inject
-  public ChatHandlerResource(ChatHandlerProvider handlers)
+  public ChatHandlerResource(ChatHandlerProvider handlers, Properties properties)
   {
     this.handlers = handlers;
+    this.taskId = properties.getProperty(TASK_ID_PROPERTY);
   }
 
   @Path("/chat/{id}")
-  public Object doTaskChat(
-      @PathParam("id") String handlerId
-  )
+  public Object doTaskChat(@PathParam("id") String handlerId, @Context HttpHeaders headers)
   {
+    if (taskId != null) {
+      List<String> requestTaskId = headers.getRequestHeader(TASK_ID_HEADER);
+      if (requestTaskId != null && !requestTaskId.contains(taskId)) {
+        return null;
+      }
+    }
+
     final Optional<ChatHandler> handler = handlers.get(handlerId);
 
     if (handler.isPresent()) {
       return handler.get();
-    } else {
-      return Response.status(Response.Status.NOT_FOUND).build();
```

Comment:
This `return Response()` is a big troll; the code **does** return a 404 to the caller, but not for the reason you would think. It returns a 404 because there's no method in the returned object annotated with a @Path param so the request doesn't get matched to any endpoint and thus the Jetty framework returns a 404. The Response object itself is not respected; in other words, you can't just set a response header on the object and have it propagate back to the caller.

Hence, I think the options are you'd either have to return a dummy object that would match any path so you would have access to the HTTP response, or you'd have to do it on the servlet filter level, which is what this PR does. I changed this to just return null since it does the exact same thing and might save someone else from being trolled in the future. :P